sidekiq_with_redis_test.rb

```diff
@@ -0,0 +1,72 @@
+# This file is distributed under New Relic's license terms.
+# See https://github.com/newrelic/newrelic-ruby-agent/blob/main/LICENSE for complete details.
+# frozen_string_literal: true
+
+require 'bundler/inline'
+
+class SidekiqWithRedisTest < MiniTest::Test
+  #
+  # Regression test for
+  #   https://github.com/newrelic/newrelic-ruby-agent/issues/1639
+  #
+  # PR 1611 introduced a new `call_pipelined_with_tracing` method to
+  # `lib/new_relic/agent/instrumentation/redis/instrumentation.rb`. That method
+  # is defined on all instrumented instances of RedisClient (defined by the
+  # `redis-client` gem) when v5.0+ of the `redis` gem (which requires
+  # `redis-client`) is used. Originally the new method performed a
+  # `client.db` call on `self` to get at the client's configured database value.
+  # At the time, `client` was assumed to always return an instance of
+  # `Redis::Client` which is a class defined by the `redis` gem. This
+  # `Redis::Client` class has a `db` helper method that exposes the configured
+  # database value.
+  #
+  # With Issue 1639 it was discovered that `client` could return an instance
+  # of `RedisClient` (defined by the `redis-client` gem) instead of
+  # `Redis::Client` (defined by the `redis` gem). The original `client.db` call
+  # was updated to read `client.config.db` instead. This approach is known to
+  # work for instances of `RedisClient` and `Redis::Client`, as
+  # `Redis::Client#db` is just a helper method that calls `config.db`.
+  #
+  # This test reproduces the problem by placing an instance of `RedisClient` in
+  # scope when the `call_pipelined_with_tracing` method calls `client`, and
+  # confirms that the instrumentation no longer (and never again) errors out.
+  #
+  # NOTE: Because Sidekiq v7.0+ can use `redis-client` without `redis`, this
+  #       test brings in the `redis` gem directly via `bundler/inline`
+  def test_redis_client_pipelined_calls_work
+    skip unless run_these_tests?
+
+    gemfile do
+      source 'https://rubygems.org'
+
+      gem 'redis'
+    end
+
+    require 'sidekiq'
+    require 'newrelic_rpm'
+
+    conn = Sidekiq::RedisConnection.create
+    key = 'pineapple'
+    value = 'carrot'
+    result = nil
+
+    conn.with do |c|
+      c._client.pipelined do |p|
+        p.call_v([:set, key, value])
+      end
+      result = c._client.call(:get, key)
+    end
+
+    assert_equal value, result
+  end
+
+  private
+
+  def run_these_tests?
+    # these tests aren't impacted by Ruby version
+    return false unless RUBY_VERSION >= '3.1.2'
```

Comment:
Im confused by the comment above this. How come we're checking the ruby version if the tests aren't impacted by the ruby version?

Reply:
Without this check, every single Ruby being tested by the CI will perform the test. Given that we're doing an inline bundling of `redis` and that the test won't manifest itself differently with a different Ruby version, this check exists to ensure that we perform as few runs of this test as possible.

I'm very open to changing or removing this check, and/or the comment above it. Give it some thought and let me know what you'd like this bit of the test to look like.

Reply:
ahh I see, so it's not really about the ruby version so much as it's just about running it just once on the CI to save time. And yeah, I would appreciate it if the comment was a little more specific about why it's doing this, since this isn't something we typically do. thank you for explaining!

Reply:
I gave it a rework with [516dd6f](https://github.com/newrelic/newrelic-ruby-agent/pull/1640/commits/516dd6f4603da7609cc2b476f95b9892ce92eac4)